README.md
# tartiflette-plugin-isodate

<a href="https://buddy.works"><img src="https://assets.buddy.works/automated-dark.svg" alt="Buddy.Works logo."></img></a>

[![buddy pipeline](https://app.buddy.works/benriggleman/tartiflette-plugin-isodate/pipelines/pipeline/208276/badge.svg?token=ff05a3fb6bb08b48350b4170e0c447aa3ccc198abbddd48c222205c3c61a7cff "buddy pipeline")](https://app.buddy.works/benriggleman/tartiflette-plugin-isodate/pipelines/pipeline/208276)

ISO Date Format Directive for Tartiflette

## TOC
- [Overview](#overview)
- [Installation](#install)
- [Usage](#usage)
    - [Options](#usage-options)
        - [microseconds](#usage-options-microseconds)
        - [timezone](#usage-options-timezone)
        - [utc](#usage-options-utc)


## [Overview](#overview)

The `tartiflette-plugin-isodatenow` plugin adds an `@isoDateNow` directive to [tartiflette](https://github.com/tartiflette/tartiflette).  It sets the field that it is on with the current time in [ISO 8601 format](https://en.wikipedia.org/wiki/ISO_8601).
It is primarly intended for use in mutations as the directive overrides any value sent though it can be used in queries where a `createdAt` or response timestamp is required.

## [Installation](#install)

To install with [poetry](https://poetry.eustace.io):

```sh
$ poetry add tartiflette-plugin-isodate
```

To install with pip:

```sh
$ pip install tartiflette-plugin-isodate
```

## [Usage](#usage)

```graphql
type Example {
    createdAt: String @isoDateNow
}
```

Querying `createdAt` would return the following:

```json
{
    "data": {
        "example": {
            "createdAt": "2019-09-04T13:49:12.585158+00:00"
        }
    }
}
```


### [Options](#usage-options)

The `@isoDateNow` also takes the following optional arguments:

#### [@isoDateNow(microseconds: false)](#usage-options-microseconds)

Returns date and time _without_ microseconds.

```graphql
type Example {
    createdAt: String @isoDateNow(microseconds: false)
}
```

Querying `createdAt` would return the following:

```json
{
    "data": {
        "example": {
            "createdAt": "2019-09-04T13:49:12+00:00"
        }
    }
}
```

#### [@isoDateNow(timezone: false)](#usage-options-timezone)

Returns date and time _without_ timezone.

```graphql
type Example {
    createdAt: String @isoDateNow(timezone: false)
}
```

Querying `createdAt` would return the following:

```json
{
    "data": {
        "example": {
            "createdAt": "2019-09-04T13:49:12.585158"
        }
    }
}
```

#### [@isoDateNow(utc: false)](#usage-options-utc)

Returns date and time in the _local_ timezone.

```graphql
type Example {
    createdAt: String @isoDateNow(utc: false)
}
```

Querying `createdAt` would return the following:

```json
{
    "data": {
        "example": {
            "createdAt": "2019-09-04T09:49:12.585158-04:00"
        }
    }
}
```


The arguments can be used in any combination and will return an [ISO8601 compliant date](https://en.wikipedia.org/wiki/ISO_8601).

For example settings `microseconds` to `false` and `timezone` to `true` would yield: 2019-09-04T13:49:12+00:00.

Possible combinations:

- `@isoDateNow` >> "2019-09-04T13:49:12.585158+00:00"
- `@isoDateNow(timezone: false)` >> "2019-09-04T13:52:43.476260"
- `@isoDateNow(microseconds: false)` >> "2019-09-04T13:50:02+00:00"
- `@isoDateNow(microseconds: false, timezone: false)` >> "2019-09-04T13:53:31"

The time can also be set to the `local` time by setting `utc` to `false`.  

`@isoDateNow(utc: false)` >> "2019-09-04T09:50:02+00:00"

Using the `@isoDateNow` directive will override any value sent.

CHANGELOG.md
# tartiflette-plugin-isodatenow

## Released Versions

### [0.2.0]
- Package cleanup

#### Fixed

#### Changed
- Package cleanup

#### Added

### [0.2.0]
- Renamed `isoDate` to `isoDateNow`

#### Fixed

#### Changed
- Renamed `isoDate` to `isoDateNow`

#### Added

### [0.1.4]
- Initial release

#### Fixed

#### Changed

#### Added

tests/unit/test_isodate.py
import pytest


@pytest.mark.asyncio
async def test_isodate_bake():
    from tartiflette_plugin_isodatenow import bake, _SDL

    assert await bake("a", {}) == _SDL


tests/functional/test_isodate_utc.py
from unittest.mock import Mock

import asyncio
import datetime
import pytest

from freezegun import freeze_time
from tartiflette import Resolver, create_engine


@freeze_time("2019-09-04T13:49:12.585158", tz_offset=0)
@pytest.mark.asyncio
async def test_isodate_defaults():
    logger = Mock()
    logger.debug = Mock()

    @Resolver("Query.createdAt", schema_name="test_isodate_defaults_utc")
    async def created_at_resolver(*_args, **_kwargs):
        return {}

    @Resolver("ISODate.date", schema_name="test_isodate_defaults_utc")
    async def created_at_date_resolver(*_args, **_kwargs):
        await asyncio.sleep(0.100)
        return "OVERWRITTEN"

    engine = await create_engine(
        sdl="""
        type ISODate {
            date: String @isoDateNow
        }
        type Query {
            createdAt: ISODate
        }
        """,
        modules=[
            {
                "name": "tartiflette_plugin_isodatenow",
                "config": {}
            }
        ],
        schema_name="test_isodate_defaults_utc",
    )
    
    assert await engine.execute("query a { createdAt { date } } ") == {
        "data": {"createdAt": {"date": "2019-09-04T13:49:12.585158+00:00"}}
    }


@freeze_time("2019-09-04T13:49:12", tz_offset=0)
@pytest.mark.asyncio
async def test_isodate_no_microseconds():
    logger = Mock()
    logger.debug = Mock()

    @Resolver("Query.createdAt", schema_name="test_isodate_no_microseconds_utc")
    async def created_at_resolver(*_args, **_kwargs):
        return {}

    @Resolver("ISODate.date", schema_name="test_isodate_no_microseconds_utc")
    async def created_at_date_resolver(*_args, **_kwargs):
        await asyncio.sleep(0.100)
        return "OVERWRITTEN"

    engine = await create_engine(
        sdl="""
        type ISODate {
            date: String @isoDateNow(microseconds: false)
        }
        type Query {
            createdAt: ISODate
        }
        """,
        modules=[
            {
                "name": "tartiflette_plugin_isodatenow",
                "config": {},
            }
        ],
        schema_name="test_isodate_no_microseconds_utc",
    )

    assert await engine.execute("query a { createdAt { date } } ") == {
        "data": {"createdAt": {"date": "2019-09-04T13:49:12+00:00"}}
    }


@freeze_time("2019-09-04T13:49:12.585158")
@pytest.mark.asyncio
async def test_isodate_no_timezone():
    logger = Mock()
    logger.debug = Mock()

    @Resolver("Query.createdAt", schema_name="test_isodate_no_timzone_utc")
    async def created_at_resolver(*_args, **_kwargs):
        return {}

    @Resolver("ISODate.date", schema_name="test_isodate_no_timzone_utc")
    async def created_at_date_resolver(*_args, **_kwargs):
        await asyncio.sleep(0.100)
        return "OVERWRITTEN"

    engine = await create_engine(
        sdl="""
        type ISODate {
            date: String @isoDateNow(timezone: false)
        }
        type Query {
            createdAt: ISODate
        }
        """,
        modules=[
            {
                "name": "tartiflette_plugin_isodatenow",
                "config": {}
            }
        ],
        schema_name="test_isodate_no_timzone_utc",
    )
    
    assert await engine.execute("query a { createdAt { date } } ") == {
        "data": {"createdAt": {"date": "2019-09-04T13:49:12.585158"}}
    }


@freeze_time("2019-09-04T13:49:12", tz_offset=0)
@pytest.mark.asyncio
async def test_isodate_no_microseconds_timezone():
    logger = Mock()
    logger.debug = Mock()

    @Resolver("Query.createdAt", schema_name="test_isodate_no_microseconds_timezone_utc")
    async def created_at_resolver(*_args, **_kwargs):
        return {}

    @Resolver("ISODate.date", schema_name="test_isodate_no_microseconds_timezone_utc")
    async def created_at_date_resolver(*_args, **_kwargs):
        await asyncio.sleep(0.100)
        return "OVERWRITTEN"

    engine = await create_engine(
        sdl="""
        type ISODate {
            date: String @isoDateNow(microseconds: false, timezone: false)
        }
        type Query {
            createdAt: ISODate
        }
        """,
        modules=[
            {
                "name": "tartiflette_plugin_isodatenow",
                "config": {},
            }
        ],
        schema_name="test_isodate_no_microseconds_timezone_utc",
    )

    assert await engine.execute("query a { createdAt { date } } ") == {
        "data": {"createdAt": {"date": "2019-09-04T13:49:12"}}
    }


tests/functional/test_isodate_local.py
from unittest.mock import Mock, patch

import asyncio
import datetime
import pytest

from freezegun import freeze_time
from tartiflette import Resolver, create_engine


@pytest.fixture
def mock_time_altzone(autouse=True):
    with patch("tartiflette_plugin_isodatenow.time") as mocked:
        mocked.altzone = 14400
        yield mocked


@freeze_time("2019-09-04T13:49:12.585158", tz_offset=-datetime.timedelta(hours=4))
@pytest.mark.asyncio
async def test_isodate_defaults(mock_time_altzone):
    logger = Mock()
    logger.debug = Mock()

    @Resolver("Query.createdAt", schema_name="test_isodate_defaults")
    async def created_at_resolver(*_args, **_kwargs):
        return {}

    @Resolver("ISODate.date", schema_name="test_isodate_defaults")
    async def created_at_date_resolver(*_args, **_kwargs):
        await asyncio.sleep(0.100)
        return "OVERWRITTEN"

    engine = await create_engine(
        sdl="""
        type ISODate {
            date: String @isoDateNow(utc: false)
        }
        type Query {
            createdAt: ISODate
        }
        """,
        modules=[
            {
                "name": "tartiflette_plugin_isodatenow",
                "config": {}
            }
        ],
        schema_name="test_isodate_defaults",
    )
    
    assert await engine.execute("query a { createdAt { date } } ") == {
        "data": {"createdAt": {"date": "2019-09-04T09:49:12.585158-04:00"}}
    }


@freeze_time("2019-09-04T13:49:12", tz_offset=-datetime.timedelta(hours=4))
@pytest.mark.asyncio
async def test_isodate_no_microseconds(mock_time_altzone):
    logger = Mock()
    logger.debug = Mock()

    @Resolver("Query.createdAt", schema_name="test_isodate_no_microseconds")
    async def created_at_resolver(*_args, **_kwargs):
        return {}

    @Resolver("ISODate.date", schema_name="test_isodate_no_microseconds")
    async def created_at_date_resolver(*_args, **_kwargs):
        await asyncio.sleep(0.100)
        return "OVERWRITTEN"

    engine = await create_engine(
        sdl="""
        type ISODate {
            date: String @isoDateNow(microseconds: false, utc: false)
        }
        type Query {
            createdAt: ISODate
        }
        """,
        modules=[
            {
                "name": "tartiflette_plugin_isodatenow",
                "config": {},
            }
        ],
        schema_name="test_isodate_no_microseconds",
    )

    assert await engine.execute("query a { createdAt { date } } ") == {
        "data": {"createdAt": {"date": "2019-09-04T09:49:12-04:00"}}
    }


@freeze_time("2019-09-04T13:49:12.585158", tz_offset=-datetime.timedelta(hours=4))
@pytest.mark.asyncio
async def test_isodate_no_timezone():
    logger = Mock()
    logger.debug = Mock()

    @Resolver("Query.createdAt", schema_name="test_isodate_no_timzone")
    async def created_at_resolver(*_args, **_kwargs):
        return {}

    @Resolver("ISODate.date", schema_name="test_isodate_no_timzone")
    async def created_at_date_resolver(*_args, **_kwargs):
        await asyncio.sleep(0.100)
        return "OVERWRITTEN"

    engine = await create_engine(
        sdl="""
        type ISODate {
            date: String @isoDateNow(timezone: false, utc: false)
        }
        type Query {
            createdAt: ISODate
        }
        """,
        modules=[
            {
                "name": "tartiflette_plugin_isodatenow",
                "config": {}
            }
        ],
        schema_name="test_isodate_no_timzone",
    )
    
    assert await engine.execute("query a { createdAt { date } } ") == {
        "data": {"createdAt": {"date": "2019-09-04T09:49:12.585158"}}
    }


@freeze_time("2019-09-04T13:49:12", tz_offset=-datetime.timedelta(hours=4))
@pytest.mark.asyncio
async def test_isodate_no_microseconds_timezone():
    logger = Mock()
    logger.debug = Mock()

    @Resolver("Query.createdAt", schema_name="test_isodate_no_microseconds_timezone")
    async def created_at_resolver(*_args, **_kwargs):
        return {}

    @Resolver("ISODate.date", schema_name="test_isodate_no_microseconds_timezone")
    async def created_at_date_resolver(*_args, **_kwargs):
        await asyncio.sleep(0.100)
        return "OVERWRITTEN"

    engine = await create_engine(
        sdl="""
        type ISODate {
            date: String @isoDateNow(microseconds: false, timezone: false, utc: false)
        }
        type Query {
            createdAt: ISODate
        }
        """,
        modules=[
            {
                "name": "tartiflette_plugin_isodatenow",
                "config": {},
            }
        ],
        schema_name="test_isodate_no_microseconds_timezone",
    )

    assert await engine.execute("query a { createdAt { date } } ") == {
        "data": {"createdAt": {"date": "2019-09-04T09:49:12"}}
    }


tartiflette_plugin_isodate/__init__.py
import logging
import time
import datetime

from tartiflette import Directive, Resolver, Scalar

_SDL = """
directive @isoDate(microseconds: Boolean! = true, timezone: Boolean! = true, utc: Boolean! = true) on FIELD_DEFINITION | FIELD
"""


class ISODate:
    def __init__(self, config):
        self._logger = config.get("logger", logging.getLogger(__name__))

    async def on_field_execution(
        self, directive_args, next_resolver, parent_result, args, ctx, info
    ):
        results = await next_resolver(parent_result, args, ctx, info)
        _time = await self._get_time(directive_args["utc"])
        _timezone = await self._get_timezone(directive_args["utc"])

        if directive_args["microseconds"] and directive_args["timezone"]:
            # microsecond = True and timezone = True: '2019-09-04T13:49:12.585158+00:00'
            results = _time.replace(tzinfo=_timezone).isoformat()
        elif directive_args["microseconds"] and not directive_args["timezone"]:
            # microsecond = True and timezone = False: '2019-09-04T13:52:43.476260'
            results = _time.isoformat()
        elif not directive_args["microseconds"] and directive_args["timezone"]:
            # microsecond = False and timezone = True: '2019-09-04T13:50:02+00:00'
            results = _time.replace(tzinfo=_timezone).isoformat()
        else:
            # microsecond = False and timezone = False: '2019-09-04T13:53:31'
            results = _time.replace(microsecond=0).isoformat()

        return results

    async def _get_time(self, _utc):
        if _utc:
            return datetime.datetime.utcnow()
        return datetime.datetime.now()

    async def _get_timezone(self, _utc):
        if _utc:
            return datetime.timezone.utc
        return datetime.timezone(offset=datetime.timedelta(seconds=-time.altzone))


async def bake(schema_name, config):
    Directive(name="isoDate", schema_name=schema_name)(ISODate(config))
    return _SDL
